System.Net.WebProxy Bypass regex list optimization

onehourlate opened the pull request:
Added change tracking to BypassArrayList to avoid reconstructing the bypass regex list at every call of IsBypassed

Fix #73703

Commit:
System.Net.WebProxy Bypass regex list optimization

Added change tracking to BypassArrayList to avoid reconstructing the bypass regex list at every call of IsBypassed

Fix #73703

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73803) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73803) before we can accept your contribution.<br/><br/>:x: onehourlate [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73803)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=73803) it.</sub>

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -11,7 +11,134 @@ namespace System.Net
+
+            public bool IsDirty { get; private set; }
+
+            public bool CheckDirty()
```
Nit: CheckDirty => GetAndResetIsDirty

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -11,7 +11,134 @@ namespace System.Net
+
+            public override object? this[int index]
+            {
+                get
+                {
+                    return base[index];
+                }
```
```suggestion
                get => base[index];
```

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -11,7 +11,134 @@ namespace System.Net
+
+            public bool CheckDirty()
+            {
+                var dirty = IsDirty;
```
```suggestion
                bool dirty = IsDirty;
```

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -11,7 +11,134 @@ namespace System.Net
     public partial class WebProxy : IWebProxy, ISerializable
     {
-        private ArrayList? _bypassList;
+
```
```suggestion
```

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -11,7 +11,134 @@ namespace System.Net
     {
-        private ArrayList? _bypassList;
+
+        private sealed class DirtyTrackingArrayList : ArrayList
```
Please move this nested type to the end rather than the beginning of WebProxy.

stephentoub replied:
Also, let's call it ChangeTrackingArrayList instead of DirtyTrackingArrayList, use IsChanged instead of IsDirty, etc.

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -121,13 +248,33 @@ public bool UseDefaultCredentials
 
-        private void UpdateRegexList(bool canThrow)
+
+        //Replaces calls to UpdateRegexList wich took a 'canThrow' flag, set to false, so we just swallow any exception.
```
```suggestion
        /// <summary>Replaces calls to UpdateRegexList wich took a 'canThrow' flag, set to false, so we just swallow any exception.</summary>
```

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -121,13 +248,33 @@ public bool UseDefaultCredentials
+            if (bypassList != null)
             {
-                if (bypassList != null && bypassList.Count > 0)
+                bypassList.CheckDirty();
```
What's the purpose of calling CheckDirty if the return value is ignored?  If the goal here is to just set IsDirty to false, it'd be better to do that explicitly.  And at that point, we don't need CheckDirty... the other call site can similarly just access the getter, and then if it's true and we enter that if block, set IsDirty to false.

onehourlate replied:
Yes, CheckDirty is not really a good name.

The initial idea was to avoid having to do a 
if (a.IsDirty) {
 // do some stuff
 a.IsDirty = false;
}

and rather simply do

if (a.CheckDirty()) {
  // do some stuff
}
Plus I though the dirty flag should only be set to true by the class, not by an external caller, so I didn't want to expose a setter for IsDirty.

But turns out, that's not it's not really used that way...


And yes, the reason it's called in the UpdateRegexList is to reset the dirty status to false for when the UpdateRegexList is called directly in the constructor and the BypassList setter. In that case, we don't care about the return value.

But if it's clearer, we can do

IsChanged {get;}
and void ResetIsChanged();

stephentoub replied:
Right. And at that point it can just be an IsChanged with a getter and setter.

onehourlate replied:
I committed a second set of changes based on your general feedback.
But I did IsChanged  {get;} / void ResetIsChanged()
I'm still not too fond of beeing able to do a.IsChanged = true; from an external caller.

stephentoub replied:
> I'm still not too fond of beeing able to do a.IsChanged = true; from an external caller. (even if this this all completely private stuff)

Why?  `ResetIsChanged()` is equivalent to setting `IsChanged` to false, and there are a myriad of other ways to set it to true, e.g. `a[0] = a[0];`, so not exposing `IsChanged` because of concerns around someone doing `IsChanged = true;` isn't preventing anything.  Honestly my inclination would be that a property is even too much here and it should just be an internal `_isChanged` field.  This type isn't even available to the rest of the assembly, nevermind public... it's a private implementation detail of WebProxy, and in that regard is just like any other private member, whose shape is there purely to serve the needs of the type it's private in.  We shouldn't spend time or energy on polishing the shape of these APIs beyond what's helpful in maintainability.  And FWIW, from my perspective the most maintainable thing is for there to be a single entity that's used to both get and set the flag.

onehourlate commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -11,7 +11,134 @@ namespace System.Net
+
+        private sealed class DirtyTrackingArrayList : ArrayList
+        {
+
+            public DirtyTrackingArrayList()
+            {
+
+            }
+
```
hum.

That's needed by

 public ArrayList BypassArrayList => _bypassList ??= new DirtyTrackingArrayList();

stephentoub replied:
If you don't add any ctors to a type, the C# compiler adds a public parameterless one by default.

onehourlate replied:
yes. but there _are_ other constructors defined.

stephentoub replied:
Ok, I didn't see the other constructors.  If they're all needed, that's fine, then.

Commit:
Cosmetic changes.

DirtyTrackingArrayList => ChangeTrackingArrayList.
IsDirty => IsChanged
Inner class moved to end
Ditched weird bool CheckDirty() with side effect in favor of void ResetIsChanged()

stephentoub commented on src/libraries/System.Net.WebProxy/tests/WebProxyTest.cs:
```diff
@@ -132,6 +132,27 @@ public static void WebProxy_InvalidBypassUrl_AddedDirectlyToList_SilentlyEaten()
+        {
+            var p = new WebProxy("http://microsoft.com", false, new string[] { "foobar" });
+
+            //not sure what the policy is for reflexion in unit tests...
```
We try to avoid it.

I don't think we need to test that instance isn't being regenerated.  I'd rather see tests that all possible updates to the ArrayList are reflected on subsequent calls, e.g. you've added explicit code to set IsChanged when regexes are added/removed/inserted/changed/etc... we should have tests then validating that all possible mutations on the returned ArrayList are reflected in subsequent IsBypassed calls.

stephentoub commented on src/libraries/System.Net.WebProxy/src/System/Net/WebProxy.cs:
```diff
@@ -121,13 +121,33 @@ public bool UseDefaultCredentials
 
-        private void UpdateRegexList(bool canThrow)
+
+        /// <summary>Replaces calls to UpdateRegexList which took a 'canThrow' flag, set to false, so we just swallow any exception.</summary>
```
```suggestion
        /// <summary>Invokes <see cref="UpdateRegexList"/> and ignores any exceptions that result.</summary>
```

stephentoub reviewed (commented):
Other than the comments, the change generally LGTM.  Thanks for working on this.

stephentoub committed:
Address PR feedback

stephentoub commented:
As we have a deadline of tomorrow, I've addressed the open feedback and pushed a new commit.  Thanks.

onehourlate commented:
Ok.
Sorry, it wasn't clear to me you were waiting for more changes...
The fact that I didn't quite agree with the suggested changes didn't help ;)

I should have added all the possible mutation tests cases, tough : that's obviously a plus.

I don't like reflection in Unit Tests either, but I would have kept that test : with the current unit tests, the initial code would pass. I thought it was a net positive (despite the use of reflexion), to have unit tests that failed for the original code.

Anyway.
To be clear : do I still have something to do for this PR, or is it now going its merry way ?

stephentoub commented:
> Sorry

No problem.

> with the current unit tests, the initial code would pass

There are literally thousands of places things are cached around the core libraries, and except where there's a visible behavioral impact, we don't have unit tests for it. If you'd like to submit a PR to add a perf test for this to dotnet/performance, that'd be welcome.

> do I still have something to do for this PR

Nope, thanks.

stephentoub commented:
Thank you, @onehourlate.
